src/components/EditProfile.js

```diff
@@ -0,0 +1,233 @@
+    }
+
+    const imageFile = e.target.files[0];
+    setSelectedFile(imageFile);
+
+    const imageName = imageFile.name.split(".");
+    setNewUserPhotoName(imageName[0]);
+  };
+
+  useEffect(() => {
+    if (!selectedFile) {
+      return;
+    }
+    const objectUrl = URL.createObjectURL(selectedFile);
+    setDisplayPhoto(objectUrl);
+    return () => URL.revokeObjectURL(objectUrl);
+  }, [selectedFile]);
+
+  const handleNewName = (e) => {
+    setNewName(e.target.value);
+  };
+  const handleNewDesc = (e) => {
+    setNewDescription(e.target.value);
+  };
+  const handleNewLink = (e) => {
+    setNewLink(e.target.value);
+  };
+
+  const back2Profile = () => {
+    const leave = confirm("user data not changed, do you want to leave?");
+    if (leave) {
+      redirect("/profile");
+    }
+  };
+
+  const cancelDataChange = () => {
+    if (!userData) {
+      alert("something went wrong, please try again later");
+
+      return;
+    }
+    const cancel = confirm("data not saved, do you want to cancel update?");
+    if (cancel) {
+      setDisplayPhoto(userData.pic);
+      setSelectedFile("");
+      setNewName(userData.name);
+      setNewLink(userData.link);
+      setNewDescription(userData.desc);
+    }
+  };
+
+  const updateUserData = async () => {
+    const userRef = doc(props.db, "user", props.uid);
+    updateDoc(userRef, {
+      name: newName,
+      link: newLink,
+      desc: newDescription,
+    });
+  };
+
+  const getUserImageUrl = async (name) => {
+    getDownloadURL(ref(storage, `profileImages/${name}`))
+      .then((url) => {
+        const userRef = doc(props.db, "user", props.uid);
+        updateDoc(userRef, {
+          pic: url,
+        });
+      })
+      .then(() => {
+        alert("profile updated!");
+        redirect("/profile");
+      });
+  };
+
+  async function uploadNewUserPhotoImage() {
+    const storageRef = ref(storage, `profileImages/${newUserPhotoName}`);
+
+    try {
+      uploadBytes(storageRef, selectedFile).then(() =>
+        getUserImageUrl(newUserPhotoName)
+      );
+    } catch (error) {
+      console.log(error);
+    }
+  }
+
+  const submitNewUserData = async () => {
+    if (!newName && !newDescription && !newLink && !selectedFile) {
+      alert("if not updating, press the leave-button on the left");
+
+      return;
+    } else if (newName || newDescription || newLink || selectedFile) {
+      if (selectedFile && !newName && !newDescription && !newLink) {
+        // only new photo
+        await uploadNewUserPhotoImage();
+      } else if ((newName || newDescription || newLink) && !selectedFile) {
+        // no new photo but other updates
+        updateUserData();
+      } else if ((newName || newDescription || newLink) && selectedFile) {
+        // new photo and other updates
+        //first upload new data to firestore
+        await updateUserData();
+        //then upload image and redirect to profile
+        await uploadNewUserPhotoImage();
+      }
+    }
```

Comment:
Please refactor your judgment logic here. It is hard to read. You could try to omit some redundant logic.